File: src/main/java/com/group/libraryapp/domain/user/loanhistory/UserLoanHistory.java
package com.group.libraryapp.domain.user.loanhistory;

import com.group.libraryapp.domain.user.User;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;

import static jakarta.persistence.GenerationType.IDENTITY;

import org.jetbrains.annotations.NotNull;

@Entity
public class UserLoanHistory {

  @Id
  @GeneratedValue(strategy = IDENTITY)
  private Long id;

  @ManyToOne
  private User user;

  private String bookName;

  private boolean isReturn;

  public UserLoanHistory() {

  }

  public UserLoanHistory(User user, String bookName, boolean isReturn) {
    this.user = user;
    this.bookName = bookName;
    this.isReturn = isReturn;
  }

  public void doReturn() {
    this.isReturn = true;
  }

  @NotNull
  public String getBookName() {
    return this.bookName;
  }

  @NotNull
  public User getUser() {
    return user;
  }
}

File: src/main/kotlin/com/group/libraryapp/domain/user/loanhistory/UserLoanHistory.kt
package com.group.libraryapp.domain.user.loanhistory

import com.group.libraryapp.domain.user.User
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.ManyToOne

@Entity
class UserLoanHistory(
    @ManyToOne
    val user: User,
    val bookName: String,
    @Enumerated(value = EnumType.STRING)
    var status: UserStatus = UserStatus.LOANED,
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,
) {
    fun doReturn() {
        this.status = UserStatus.RETURNED
    }

    companion object {
        fun fixture(
            borrower: User,
            bookName: String,
            status: UserStatus,
        ): UserLoanHistory = UserLoanHistory(borrower, bookName, status, null)
    }
}
